File: Introduction to Data Structures: Binary Tree/1. Traverse A Tree/1. Binary Tree Preorder Traversal/2. Iterative Method/BinaryTreePreorderTraversalIterative.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

 class Solution {
    
    public List<Integer> preorderTraversal(TreeNode root) {
        
        // create a list to hold the output
        List<Integer> output = new ArrayList<>();
    
        // create a stack to iteratively perform the operation
        Stack<TreeNode> stack = new Stack<>(); 
        
        // push root node to stack
        stack.push(root);
        
        while (!stack.empty()) {
            
            TreeNode node = stack.pop();
            
                if (node != null) {
                    // append value at node
                    output.add(node.val);

                    // add right right to the stack
                    stack.push(node.right);

                    // add left child to the stack
                    stack.push(node.left);
            }
        }

        return output;
    }
}

File: Introduction to Data Structures: Binary Tree/1. Traverse A Tree/1. Binary Tree Preorder Traversal/2. Iterative Method/BinaryTreePreorderTraversalIterative.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:

    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:

        # intialize a list to keep values to return
        output = []

        # initalize a stack with the root node
        stack = [root]

        # while the stack is not empty
        while stack:

            # pop the last item
            node = stack.pop()

            # if node is not None
            if node:

                # append the node value to the output
                output.append(node.val)

                # append right child to stack
                stack.append(node.right)

                # append left child to stack
                stack.append(node.left)

        return output
